File: Number of Ways to Arrive at Destination - GFG/number-of-ways-to-arrive-at-destination.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++
#define ll long long
class Solution {
  public:
    int countPaths(int n, vector<vector<int>>& roads) {
        // code here
        ll u,v,w,node,d,adjnode,edw;
        vector<pair<ll,ll>>adj[n];
        // vector<pair<ll,pair<ll,ll>>>adj;
        for(auto i:roads)
        {
            u=i[0];
            v=i[1];
            w=i[2];
            adj[u].push_back({v,w});
            adj[v].push_back({u,w});
        }
        vector<ll>dist(n,1e18),ways(n,0);
        dist[0]=0;
        ways[0]=1;
        priority_queue<pair<ll,ll>, vector<pair<ll,ll>>, greater<pair<ll,ll>>>pq;
        pq.push({0,0});
        ll mod=(ll)(1e9+7);
        while(!pq.empty()){
            d=pq.top().first;
            node=pq.top().second;
            pq.pop();
            for(auto i:adj[node])
            {
                adjnode=i.first;
                edw=i.second;
                if(d+edw<dist[adjnode])
                {
                    dist[adjnode]=d+edw;
                    pq.push({dist[adjnode],adjnode});
                    ways[adjnode]=ways[node];
                }else if(d+edw==dist[adjnode])
                {
                    ways[adjnode]=(ways[adjnode]+ways[node])%mod;
                }
            }
        }
        
        return ways[n-1]%mod;
    }
};

//{ Driver Code Starts.

int main() {
    int T;
    cin >> T;
    while (T--) {
        int n, m;
        cin >> n >> m;
        int u, v;

        vector<vector<int>> adj;

        for (int i = 0; i < m; ++i) {
            vector<int> temp;
            for (int j = 0; j < 3; ++j) {
                int x;
                cin >> x;
                temp.push_back(x);
            }
            adj.push_back(temp);
        }

        Solution obj;
        cout << obj.countPaths(n, adj) << "\n";
    }

    return 0;
}
// } Driver Code Ends

File: Number of Ways to Arrive at Destination - GFG/number-of-ways-to-arrive-at-destination.py
#User function Template for python3
#User function Template for python3

from typing import List
from collections import defaultdict as dd
import sys
from heapq import *
class Solution:
    def countPaths(self, n: int, roads: List[List[int]]) -> int:
        #Your code here
        
        graph = dd(list)
        for u,v,t in roads:
            graph[u].append([v,t])
            graph[v].append([u,t])
            
        ways = [0 for i in range(n)]
        dist = [1e10 for i in range(n)]
        mod = 10**9+7
        q = []
        heapify(q)
        q.append([0,0])
        ways[0] = 1
        dist[0] = 0
        while q:
            time,node = heappop(q)
            for adj,t in graph[node]:
                if time + t < dist[adj]:
                    ways[adj] = ways[node]
                    dist[adj] = time + t
                    heappush(q,[dist[adj],adj])
                elif time + t == dist[adj]:
                    ways[adj] = (ways[adj]+ways[node])%mod
        return [0,ways[n-1]%mod][dist[n-1]!=1e9]

#{ 
 # Driver Code Starts
#Initial Template for Python 3

class IntArray:
    def __init__(self) -> None:
        pass
    def Input(self,n):
        arr=[int(i) for i in input().strip().split()]#array input
        return arr
    def Print(self,arr):
        for i in arr:
            print(i,end=" ")
        print()


if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        n,m=map(int,input().split())
        
        
        adj=[]
        
        for i in range(m):
            tmp =[]
            x,y,z=map(int,input().split())
            tmp.append(x)
            tmp.append(y)
            tmp.append(z)
            adj.append(tmp)
            
        
        
        
       
        obj = Solution()
        res = obj.countPaths(n, adj)
        
        print(res)
        

# } Driver Code Ends